x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/stats/StatsDirectoryWrapper.java

```diff
@@ -0,0 +1,260 @@
+    }
+
+    @Override
+    public IndexInput openInput(final String name, final IOContext context) throws IOException {
+        final LiveStats stats = records.computeIfAbsent(name, n -> new LiveStats()).incrementOpenCount();
+        return new StatsIndexInputWrapper(super.openInput(name, context), stats);
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            super.close();
+        } finally {
+            records.clear();
+        }
+    }
+
+    /**
+     * Wraps an {@link IndexInput} to record stats about how it is accessed.
+     */
+    private static class StatsIndexInputWrapper extends IndexInput {
+
+        /**
+         * The IndexInput to delegate operations to
+         **/
+        private final IndexInput input;
+        /**
+         * The statistics to update on IndexInput operations
+         **/
+        private final LiveStats stats;
+        /**
+         * The last read position is kept around in order to detect (non)contiguous reads
+         **/
+        private final AtomicLong lastReadPosition;
+        /**
+         * Indicate if the IndexInput is a clone (or slice)
+         **/
+        private boolean isClone;
+
+        private StatsIndexInputWrapper(final IndexInput indexInput, final LiveStats stats) {
+            super("StatsIndexInputWrapper(" + indexInput + ")");
+            this.input = Objects.requireNonNull(indexInput);
+            this.stats = Objects.requireNonNull(stats);
+            this.lastReadPosition = new AtomicLong();
+        }
+
+        @Override
+        public long getFilePointer() {
+            return input.getFilePointer();
+        }
+
+        @Override
+        public long length() {
+            return input.length();
+        }
+
+        @Override
+        public void close() throws IOException {
+            try {
+                input.close();
+            } finally {
+                if (isClone == false) {
+                    stats.incrementCloseCount();
+                }
+            }
+        }
+
+        @Override
+        public byte readByte() throws IOException {
+            final long filePointer = getFilePointer();
+
+            final byte readByte = input.readByte();
+            if (readByte != -1) {
+                stats.incrementBytesRead(lastReadPosition.getAndUpdate(l -> getFilePointer()) == filePointer);
+            }
+            return readByte;
+        }
+
+        @Override
+        public void readBytes(byte[] b, int offset, int len) throws IOException {
+            final long filePointer = getFilePointer();
+            input.readBytes(b, offset, len);
+            stats.incrementBytesRead(lastReadPosition.getAndUpdate(l -> getFilePointer()) == filePointer, len);
+        }
+
+        @Override
+        public void seek(final long pos) throws IOException {
+            final long filePointer = getFilePointer();
+            input.seek(pos);
+            stats.incrementSeekCount(pos - filePointer);
```

Comment:
I wonder if this works with buffered index input that use `seekInternal` and `readInternal` to access the underlying  input ? The count could be misleading since it doesn't take into account the internal seek that the buffered input imposes ?

Reply:
I agree the count could be misleading. 

My initial wish was to have a filter directory that tracks stats and that could wrap any other directory, so that we could capture stats at different layers. For example, capturing stats at the top level searchable snapshot directory or capturing stats at another filter directory (that could take care of caching index input parts etc). I'm still struggling to follow this design though.

Knowing how Lucene access the top level directory is very interesting and correlate the numbers  with the internal buffered reads executed by SearchableSnapshotIndexInput is also nice to have so I followed your suggestion and moved stats into the searchable snapshot dir and input. It allows to push the `LiveStats` object further down in case we need to easily track other things.